core/trino-main/src/main/java/io/trino/execution/scheduler/BinPackingNodeAllocatorService.java

```diff
@@ -0,0 +1,435 @@
+    static final int PROCESS_PENDING_ACQUIRES_DELAY_SECONDS = 5;
+
+    private final InternalNodeManager nodeManager;
+    private final Supplier<Map<String, Optional<MemoryInfo>>> workerMemoryInfoSupplier;
+
+    private final ScheduledThreadPoolExecutor executor = new ScheduledThreadPoolExecutor(2, daemonThreadsNamed("bin-packing-node-allocator"));
+    private final AtomicBoolean started = new AtomicBoolean();
+    private final AtomicBoolean stopped = new AtomicBoolean();
+    private final Semaphore processSemaphore = new Semaphore(0);
+    private final AtomicReference<Map<String, Long>> nodePoolSizes = new AtomicReference<>(ImmutableMap.of());
+    private final boolean scheduleOnCoordinator;
+
+    private final ConcurrentMap<String, Long> allocatedMemory = new ConcurrentHashMap<>();
+    private final Deque<PendingAcquire> pendingAcquires = new ConcurrentLinkedDeque<>();
+
+    @Inject
+    public BinPackingNodeAllocatorService(
+            InternalNodeManager nodeManager,
+            ClusterMemoryManager clusterMemoryManager,
+            NodeSchedulerConfig config)
+    {
+        this(nodeManager,
+                requireNonNull(clusterMemoryManager, "clusterMemoryManager is null")::getWorkerMemoryInfo,
+                config.isIncludeCoordinator());
+    }
+
+    @VisibleForTesting
+    BinPackingNodeAllocatorService(
+            InternalNodeManager nodeManager,
+            Supplier<Map<String, Optional<MemoryInfo>>> workerMemoryInfoSupplier,
+            boolean scheduleOnCoordinator)
+    {
+        this.nodeManager = requireNonNull(nodeManager, "nodeManager is null");
+        this.workerMemoryInfoSupplier = requireNonNull(workerMemoryInfoSupplier, "workerMemoryInfoSupplier is null");
+        this.scheduleOnCoordinator = scheduleOnCoordinator;
+    }
+
+    @PostConstruct
+    public void start()
+    {
+        if (started.compareAndSet(false, true)) {
+            executor.schedule(() -> {
+                while (!stopped.get()) {
+                    try {
+                        // pending acquires are processed when node is released (semaphore is bumped) and periodically (every couple seconds)
+                        // in case node list in cluster have changed.
+                        processSemaphore.tryAcquire(PROCESS_PENDING_ACQUIRES_DELAY_SECONDS, TimeUnit.SECONDS);
+                        processSemaphore.drainPermits();
+                        processPendingAcquires();
+                    }
+                    catch (InterruptedException e) {
+                        currentThread().interrupt();
+                    }
+                    catch (Exception e) {
+                        // ignore to avoid getting unscheduled
+                        log.warn(e, "Error updating nodes");
+                    }
+                }
+            }, 0, TimeUnit.SECONDS);
+        }
+
+        refreshNodePoolSizes();
+        executor.scheduleWithFixedDelay(this::refreshNodePoolSizes, 1, 5, TimeUnit.SECONDS);
```

Comment:
I don't get it why we need a separate thread to refresh NodePoolSizes. Why not just try to get a snapshot in `processPendingAcquires`?

Reply:
I'd like to keep those separate. node pools are not expect to change often. Actually running it every 5s coudl be an overkill. While `processPendingAcquires` is called frequently and we want to keep it quick.